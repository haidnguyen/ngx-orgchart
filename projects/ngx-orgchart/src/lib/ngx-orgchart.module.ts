import { CommonModule } from '@angular/common';
import { ModuleWithProviders, NgModule } from '@angular/core';

import { NgxOrgchartComponent } from './ngx-orgchart.component';
import { NODE_HEIGHT, NODE_WIDTH } from './ngx-orgchart.token';
import { NodeComponent } from './node.component';

interface NgxOrgchartConfig {
  nodeWidth: number;
  nodeHeight: number;
}

@NgModule({
  declarations: [NgxOrgchartComponent, NodeComponent],
  imports: [CommonModule],
  exports: [NgxOrgchartComponent],
})
export class NgxOrgchartModule {
  static withConfig(config: NgxOrgchartConfig): ModuleWithProviders<NgxOrgchartModule> {
    return {
      ngModule: NgxOrgchartModule,
      providers: Array.of(
        { provide: NODE_WIDTH, useValue: config.nodeWidth },
        { provide: NODE_HEIGHT, useValue: config.nodeHeight }
      ),
    };
  }
}
