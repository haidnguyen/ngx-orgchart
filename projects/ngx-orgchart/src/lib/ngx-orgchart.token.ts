import { InjectionToken } from '@angular/core';

export const NODE_WIDTH = new InjectionToken<number>('NGX_ORGCHART_NODE_WIDTH');
export const NODE_HEIGHT = new InjectionToken<number>('NGX_ORGCHART_NODE_HEIGHT');
