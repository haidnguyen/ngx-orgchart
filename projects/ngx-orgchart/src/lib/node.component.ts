import { Component, OnInit, ViewEncapsulation, ChangeDetectionStrategy } from '@angular/core';

@Component({
  selector: 'app-node',
  template: `
    <p>
      node works!
    </p>
  `,
  styles: [
  ],
  encapsulation: ViewEncapsulation.None,
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class NodeComponent implements OnInit {

  constructor() { }

  ngOnInit(): void {
  }

}
