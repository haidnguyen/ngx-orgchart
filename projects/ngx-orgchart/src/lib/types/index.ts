export * from './OrgChartNode';
export * from './Point';
export * from './Event';
export * from './Entities';
