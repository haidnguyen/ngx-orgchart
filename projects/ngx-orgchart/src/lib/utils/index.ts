export * from './normalize';
export * from './tree';
