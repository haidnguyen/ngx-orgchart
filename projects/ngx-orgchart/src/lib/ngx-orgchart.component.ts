import { ChangeDetectionStrategy, Component, Input, TemplateRef, ViewEncapsulation } from '@angular/core';

import { NodeStore } from './node.store';
import { Entities, OrgChartNode } from './types';

@Component({
  selector: 'ngx-orgchart',
  template: `
    <ng-container *ngIf="vm$ | async as vm">
      <div [ngStyle]="{ 'width.px': vm.chartWidth }">
        <ngx-node *ngFor="let id of vm.rootIds" [nodeId]="id" [nodeTemplateRef]="nodeTemplateRef"></ngx-node>
      </div>
    </ng-container>
  `,
  styles: [],
  changeDetection: ChangeDetectionStrategy.OnPush,
  encapsulation: ViewEncapsulation.None,
  providers: [NodeStore],
})
export class NgxOrgchartComponent<T, K> {
  @Input()
  set datasource(value: Entities<OrgChartNode<T, K>>) {
    this.fromNodeStore.setEntities(value);
  }

  @Input() nodeTemplateRef!: TemplateRef<T>;

  readonly vm$ = this.fromNodeStore.select(
    this.fromNodeStore.rootIds$,
    this.fromNodeStore.chartWidth$,
    (rootIds, chartWidth) => ({ rootIds, chartWidth })
  );

  constructor(private readonly fromNodeStore: NodeStore<T, K>) {}
}
