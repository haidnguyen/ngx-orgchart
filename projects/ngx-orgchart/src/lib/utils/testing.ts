import { Provider } from '@angular/core';

import { NodeStore } from '../node.store';
import { Entities, OrgChartNode } from '../types';

export const provideMockNodeStore = <T extends OrgChartNode>(entities: Entities<T>): Provider => {
  const instance = new NodeStore(300);
  instance.setEntities(entities);
  return { provide: NodeStore, useValue: instance };
};
