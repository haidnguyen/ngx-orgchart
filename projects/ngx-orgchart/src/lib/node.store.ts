import { Inject, Injectable } from '@angular/core';
import { ComponentStore } from '@ngrx/component-store';
import update from 'immutability-helper';

import { NODE_WIDTH } from './ngx-orgchart.token';
import { Entities, OrgChartNode } from './types';
import { buildTree, calWidth } from './utils';

export interface NodeState<T, K> {
  entities: Entities<OrgChartNode<T, K>>;
  ids: string[];
}

@Injectable()
export class NodeStore<T = any, K = any> extends ComponentStore<NodeState<T, K>> {
  constructor(@Inject(NODE_WIDTH) private nodeWidth: number) {
    super({ entities: {}, ids: [] });
  }

  private readonly entities$ = this.select(state => state.entities);

  private readonly ids$ = this.select(state => state.ids);

  readonly nodes$ = this.select(this.entities$, this.ids$, (entities, ids) => ids.map(id => entities[id]));

  readonly rootIds$ = this.select(this.ids$, this.entities$, (ids, entities) =>
    ids.filter(id => entities[id].parentId === null)
  );

  readonly selectNodeById = (nodeId: OrgChartNode['id']) => this.select(this.entities$, entities => entities[nodeId]);

  readonly selectChildrenIds = (nodeId: OrgChartNode<T, K>['id']) =>
    this.select(this.entities$, this.ids$, (entities, ids) => ids.filter(id => entities[id].parentId === nodeId));

  readonly chartWidth$ = this.select(this.rootIds$, this.entities$, this.ids$, (rootIds, entities, ids) =>
    rootIds.map(rootId => calWidth(buildTree(rootId, ids, entities), this.nodeWidth)).reduce((acc, cur) => acc + cur)
  );

  readonly setEntities = this.updater((state, value: Entities<OrgChartNode<T, K>>) =>
    update(state, {
      entities: { $set: value },
      ids: { $set: Object.keys(value) },
    })
  );
}
