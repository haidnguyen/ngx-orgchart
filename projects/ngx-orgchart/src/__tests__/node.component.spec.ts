import { render } from '@testing-library/angular';

import { NodeComponent } from '../lib/node.component';
import { normalize } from '../lib/utils';
import { provideMockNodeStore } from '../lib/utils/testing';

interface TestData {
  name: string;
}

interface TestMeta {
  isEditing: boolean;
}

describe(NodeComponent.name, () => {
  it('should match snapshot', async () => {
    const { container } = await render<NodeComponent<TestData, TestMeta>>(NodeComponent, {
      providers: [
        provideMockNodeStore(
          normalize([
            {
              id: '1',
              parentId: null,
              data: {
                name: 'Root',
              },
              meta: {
                isEditing: false,
              },
            },
            {
              id: '2',
              parentId: '1',
              data: {
                name: 'Child 1',
              },
            },
            {
              id: '3',
              parentId: '1',
              data: {
                name: 'Child 2',
              },
            },
            {
              id: '4',
              parentId: '2',
              data: {
                name: 'Child 3',
              },
            },
          ])
        ),
      ],
      componentProperties: {
        nodeId: '1',
      },
    });

    expect(container).toMatchSnapshot();
  });
});
