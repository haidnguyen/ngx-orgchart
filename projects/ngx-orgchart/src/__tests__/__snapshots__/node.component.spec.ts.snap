// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NodeComponent should match snapshot 1`] = `
<div
  class="ngx-node-root"
  id="root0"
>
  
  <div
    class="ngx-node-container"
  >
    
    <div
      class="ngx-node-children"
    >
      
      <ngx-node
        class="ngx-node-root"
      >
        
        <div
          class="ngx-node-container"
        >
          
          <div
            class="ngx-node-children"
          >
            
            <ngx-node
              class="ngx-node-root"
            >
              
              <div
                class="ngx-node-container"
              >
                
                <div
                  class="ngx-node-children"
                >
                  
                </div>
              </div>
            </ngx-node>
          </div>
        </div>
      </ngx-node>
      <ngx-node
        class="ngx-node-root"
      >
        
        <div
          class="ngx-node-container"
        >
          
          <div
            class="ngx-node-children"
          >
            
          </div>
        </div>
      </ngx-node>
    </div>
  </div>
</div>
`;
